File: backend/src/main/java/com/stockanalysis/backtest/FeeSchedule.java
package com.stockanalysis.backtest;

import java.util.LinkedHashMap;
import java.util.Map;

/**
 * Commission per traded symbol, in percent per side. Rates are per symbol rather than per strategy
 * because a leverage ETF and its inverse can sit at different rates, and an ETF pre-market run
 * trades both — so one number per run would be wrong.
 *
 * <p>Keys match {@link TradeRecord#instrument()}: {@code null} for a single-dataset run, otherwise
 * {@code LEVERAGE} / {@code INVERSE}. The futures series is only read for its trend, never traded,
 * so it has no rate here.
 */
public record FeeSchedule(double singlePct, double leveragePct, double inversePct) {

    public static final String LEVERAGE = "LEVERAGE";
    public static final String INVERSE = "INVERSE";
    /** Key used in reporting for a single-dataset run, where {@code instrument} is null. */
    public static final String SINGLE = "SINGLE";

    /** The same rate everywhere — for single-dataset runs and tests. */
    public static FeeSchedule flat(double pct) {
        return new FeeSchedule(pct, pct, pct);
    }

    /** Commission-free: the price-only baseline. */
    public static FeeSchedule free() {
        return flat(0.0);
    }

    public static FeeSchedule ofEtf(double leveragePct, double inversePct) {
        return new FeeSchedule(0.0, leveragePct, inversePct);
    }

    /** Rate for a trade's instrument; unknown instruments fall back to the single-dataset rate. */
    public double rateFor(String instrument) {
        if (instrument == null) {
            return singlePct;
        }
        return switch (instrument) {
            case LEVERAGE -> leveragePct;
            case INVERSE -> inversePct;
            default -> singlePct;
        };
    }

    /** The rates that actually applied, keyed by instrument, for reporting on the result. */
    public Map<String, Double> ratesFor(Iterable<TradeRecord> trades) {
        Map<String, Double> used = new LinkedHashMap<>();
        for (TradeRecord t : trades) {
            String key = t.instrument() == null ? SINGLE : t.instrument();
            used.putIfAbsent(key, rateFor(t.instrument()));
        }
        return used;
    }
}

File: backend/src/main/kotlin/com/stockanalysis/backtest/FeeSchedule.kt
package com.stockanalysis.backtest

/**
 * 종목별 수수료율(편도 %).
 *
 * 전략이 아니라 **종목**에 붙는 이유: 레버리지 ETF와 인버스가 서로 다른 요율일 수 있고,
 * 장전 모드는 하루에 둘 중 하나를 고르므로 실행당 요율 하나로는 반드시 틀립니다.
 *
 * 키는 [TradeRecord.instrument]와 같습니다 — 단일 데이터셋 실행은 `null`, 그 외에는
 * `LEVERAGE`/`INVERSE`. 선물 시리즈는 추세를 읽을 뿐 매매하지 않으므로 요율이 없습니다.
 */
data class FeeSchedule(
    @get:JvmName("singlePct") val singlePct: Double,
    @get:JvmName("leveragePct") val leveragePct: Double,
    @get:JvmName("inversePct") val inversePct: Double
) {

    /** 거래의 instrument에 맞는 요율. 모르는 값이면 단일 데이터셋 요율로 떨어집니다. */
    fun rateFor(instrument: String?): Double = when (instrument) {
        null -> singlePct
        LEVERAGE -> leveragePct
        INVERSE -> inversePct
        else -> singlePct
    }

    /** 실제로 적용된 요율을 instrument별로 모읍니다 — 결과 화면에 보여주기 위한 것입니다. */
    fun ratesFor(trades: Iterable<TradeRecord>): Map<String, Double> {
        val used = LinkedHashMap<String, Double>()
        for (t in trades) {
            val key = t.instrument ?: SINGLE
            used.putIfAbsent(key, rateFor(t.instrument))
        }
        return used
    }

    companion object {
        const val LEVERAGE = "LEVERAGE"
        const val INVERSE = "INVERSE"

        /** 단일 데이터셋 실행에서 instrument가 null일 때 보고용으로 쓰는 키. */
        const val SINGLE = "SINGLE"

        /** 어디나 같은 요율 — 단일 데이터셋 실행과 테스트용. */
        @JvmStatic
        fun flat(pct: Double): FeeSchedule = FeeSchedule(pct, pct, pct)

        /** 수수료 없음. 가격만 보는 기준선입니다. */
        @JvmStatic
        fun free(): FeeSchedule = flat(0.0)

        @JvmStatic
        fun ofEtf(leveragePct: Double, inversePct: Double): FeeSchedule =
            FeeSchedule(0.0, leveragePct, inversePct)
    }
}
